File: packages/alpaca-time/src/range.ts
import { addTradingDays, isTradingDate } from './calendar';
import { parseDate } from './clock';
import { fail } from './error';
import {
  addUtcDays,
  datePartsToUtcDate,
  formatDateParts,
  parseDateParts,
  utcDateToDateParts,
} from './internal';
import type { DateRange, NyDateString, WeekdayCode } from './types';

function toUtcDate(date: NyDateString): Date {
  return datePartsToUtcDate(parseDateParts(date));
}

function toDateString(date: Date): NyDateString {
  return formatDateParts(utcDateToDateParts(date));
}

function mondayStart(date: Date): Date {
  const daysSinceMonday = (date.getUTCDay() + 6) % 7;
  return addUtcDays(date, -daysSinceMonday);
}

function isoWeekInfo(date: Date): { year: number; week: number } {
  const thursday = addUtcDays(date, 3 - ((date.getUTCDay() + 6) % 7));
  const isoYear = thursday.getUTCFullYear();
  const jan4 = new Date(Date.UTC(isoYear, 0, 4));
  const week1Monday = mondayStart(jan4);
  const diffDays = Math.trunc((mondayStart(date).getTime() - week1Monday.getTime()) / 86_400_000);
  return { year: isoYear, week: Math.floor(diffDays / 7) + 1 };
}

function weekdayNumber(weekday: WeekdayCode): number {
  switch (weekday) {
    case 'mon':
      return 1;
    case 'tue':
      return 2;
    case 'wed':
      return 3;
    case 'thu':
      return 4;
    case 'fri':
      return 5;
    case 'sat':
      return 6;
    case 'sun':
      return 0;
    default:
      fail('invalid_weekday', `invalid weekday: ${weekday}`);
  }
}

export function addDays(date: NyDateString, days: number): NyDateString {
  if (!Number.isInteger(days)) {
    fail('invalid_days', `days must be integer: ${days}`);
  }

  return toDateString(addUtcDays(toUtcDate(parseDate(date)), days));
}

export function dates(startDate: NyDateString, endDate: NyDateString): NyDateString[] {
  const start = toUtcDate(parseDate(startDate));
  const end = toUtcDate(parseDate(endDate));

  if (start.getTime() > end.getTime()) {
    fail('invalid_date_range', `start_date must be <= end_date: ${startDate} > ${endDate}`);
  }

  const values: NyDateString[] = [];
  for (let current = start; current.getTime() <= end.getTime(); current = addUtcDays(current, 1)) {
    values.push(toDateString(current));
  }
  return values;
}

export function tradingDates(startDate: NyDateString, endDate: NyDateString): NyDateString[] {
  return dates(startDate, endDate).filter((date) => isTradingDate(date));
}

export function nthWeekday(
  year: number,
  month: number,
  weekday: WeekdayCode,
  nth: number,
): NyDateString {
  if (!Number.isInteger(year) || !Number.isInteger(month) || month < 1 || month > 12 || !Number.isInteger(nth) || nth < 1) {
    fail('invalid_weekday_selection', `invalid nth weekday input: year=${year}, month=${month}, nth=${nth}`);
  }

  let current = new Date(Date.UTC(year, month - 1, 1));
  const targetWeekday = weekdayNumber(weekday);
  while (current.getUTCDay() !== targetWeekday) {
    current = addUtcDays(current, 1);
  }
  current = addUtcDays(current, (nth - 1) * 7);
  if (current.getUTCMonth() !== month - 1) {
    fail('invalid_weekday_selection', `weekday ${weekday} #${nth} does not exist in ${year}-${month}`);
  }
  return toDateString(current);
}

export function isLastTradingDateOfWeek(date: string): boolean {
  try {
    const canonicalDate = parseDate(date);
    if (!isTradingDate(canonicalDate)) {
      return false;
    }

    const current = toUtcDate(canonicalDate);
    const next = toUtcDate(addTradingDays(canonicalDate, 1));
    return mondayStart(current).getTime() !== mondayStart(next).getTime();
  } catch {
    return false;
  }
}

export function weeklyLastTradingDates(startDate: string, endDate: string): NyDateString[] {
  return tradingDates(startDate, endDate).filter((date) => isLastTradingDateOfWeek(date));
}

export function lastTradingDateOfWeek(date: NyDateString): NyDateString {
  const week = calendarWeekRange(date);
  const values = weeklyLastTradingDates(week.start_date, week.end_date);
  if (values.length === 0) {
    fail('missing_last_trading_date_of_week', `no trading day found for week containing ${date}`);
  }
  return values[0];
}

export function calendarWeekRange(date: string): DateRange {
  const current = toUtcDate(parseDate(date));
  const start = mondayStart(current);
  const end = addUtcDays(start, 6);
  return {
    start_date: toDateString(start),
    end_date: toDateString(end),
  };
}

export function isoWeekRange(year: number, week: number): DateRange {
  if (!Number.isInteger(year) || !Number.isInteger(week) || week < 1 || week > 53) {
    fail('invalid_iso_week', `invalid iso week: year=${year}, week=${week}`);
  }

  const jan4 = new Date(Date.UTC(year, 0, 4));
  const week1Monday = mondayStart(jan4);
  const start = addUtcDays(week1Monday, (week - 1) * 7);
  const info = isoWeekInfo(start);
  if (info.year !== year || info.week !== week) {
    fail('invalid_iso_week', `invalid iso week: year=${year}, week=${week}`);
  }

  return {
    start_date: toDateString(start),
    end_date: toDateString(addUtcDays(start, 6)),
  };
}

File: crates/alpaca-time/src/range.rs
use chrono::{Datelike, Duration, NaiveDate, Weekday};

use crate::calendar::{add_trading_days, is_trading_date_naive};
use crate::clock::parse_naive_date;
use crate::error::{TimeError, TimeResult};
use crate::types::DateRange;

fn format_date(date: NaiveDate) -> String {
    date.format("%Y-%m-%d").to_string()
}

fn parse_weekday_code(weekday: &str) -> TimeResult<Weekday> {
    match weekday {
        "mon" => Ok(Weekday::Mon),
        "tue" => Ok(Weekday::Tue),
        "wed" => Ok(Weekday::Wed),
        "thu" => Ok(Weekday::Thu),
        "fri" => Ok(Weekday::Fri),
        "sat" => Ok(Weekday::Sat),
        "sun" => Ok(Weekday::Sun),
        _ => Err(TimeError::new(
            "invalid_weekday",
            format!("invalid weekday: {weekday}"),
        )),
    }
}

pub fn add_days(date: &str, days: i32) -> TimeResult<String> {
    let date = parse_naive_date(date)?;
    Ok(format_date(date + Duration::days(days as i64)))
}

pub fn dates(start_date: &str, end_date: &str) -> TimeResult<Vec<String>> {
    let start = parse_naive_date(start_date)?;
    let end = parse_naive_date(end_date)?;

    if start > end {
        return Err(TimeError::new(
            "invalid_date_range",
            format!("start_date must be <= end_date: {start_date} > {end_date}"),
        ));
    }

    let mut values = Vec::new();
    let mut current = start;
    while current <= end {
        values.push(format_date(current));
        current = current
            .succ_opt()
            .ok_or_else(|| TimeError::new("date_overflow", "date overflow while building list"))?;
    }

    Ok(values)
}

pub fn trading_dates(start_date: &str, end_date: &str) -> TimeResult<Vec<String>> {
    Ok(dates(start_date, end_date)?
        .into_iter()
        .filter(|date| parse_naive_date(date).map(is_trading_date_naive).unwrap_or(false))
        .collect())
}

pub fn nth_weekday(year: i32, month: u32, weekday: &str, nth: u32) -> TimeResult<String> {
    if !(1..=12).contains(&month) || nth == 0 {
        return Err(TimeError::new(
            "invalid_weekday_selection",
            format!("invalid nth weekday input: year={year}, month={month}, nth={nth}"),
        ));
    }

    let target_weekday = parse_weekday_code(weekday)?;
    let mut current = NaiveDate::from_ymd_opt(year, month, 1).ok_or_else(|| {
        TimeError::new(
            "invalid_weekday_selection",
            format!("invalid nth weekday input: year={year}, month={month}, nth={nth}"),
        )
    })?;

    while current.weekday() != target_weekday {
        current = current
            .succ_opt()
            .ok_or_else(|| TimeError::new("date_overflow", "date overflow while selecting weekday"))?;
    }

    current += Duration::days(((nth - 1) * 7) as i64);
    if current.month() != month {
        return Err(TimeError::new(
            "invalid_weekday_selection",
            format!("weekday {weekday} #{nth} does not exist in {year}-{month:02}"),
        ));
    }

    Ok(format_date(current))
}

pub fn is_last_trading_date_of_week(date: &str) -> bool {
    let Ok(date) = parse_naive_date(date) else {
        return false;
    };
    if !is_trading_date_naive(date) {
        return false;
    }

    let Ok(next_trading) = add_trading_days(&format_date(date), 1)
        .and_then(|value| parse_naive_date(&value)) else {
        return false;
    };
    let current_week_start = date - Duration::days(date.weekday().num_days_from_monday() as i64);
    let next_week_start =
        next_trading - Duration::days(next_trading.weekday().num_days_from_monday() as i64);
    current_week_start != next_week_start
}

pub fn weekly_last_trading_dates(start_date: &str, end_date: &str) -> TimeResult<Vec<String>> {
    let mut values = Vec::new();
    for date in trading_dates(start_date, end_date)? {
        if is_last_trading_date_of_week(&date) {
            values.push(date);
        }
    }
    Ok(values)
}

pub fn last_trading_date_of_week(date: &str) -> TimeResult<String> {
    let week = calendar_week_range(date)?;
    weekly_last_trading_dates(&week.start_date, &week.end_date)?
        .into_iter()
        .next()
        .ok_or_else(|| {
            TimeError::new(
                "missing_last_trading_date_of_week",
                format!("no trading day found for week containing {date}"),
            )
        })
}

pub fn calendar_week_range(date: &str) -> TimeResult<DateRange> {
    let date = parse_naive_date(date)?;
    let start = date - Duration::days(date.weekday().num_days_from_monday() as i64);
    let end = start + Duration::days(6);
    Ok(DateRange {
        start_date: format_date(start),
        end_date: format_date(end),
    })
}

pub fn iso_week_range(year: i32, week: u32) -> TimeResult<DateRange> {
    let start = NaiveDate::from_isoywd_opt(year, week, Weekday::Mon).ok_or_else(|| {
        TimeError::new(
            "invalid_iso_week",
            format!("invalid iso week: year={year}, week={week}"),
        )
    })?;
    let end = start + Duration::days(6);
    Ok(DateRange {
        start_date: format_date(start),
        end_date: format_date(end),
    })
}
